### PR TITLE
Ask, conversation, conversations commands

### PR DESCRIPTION
**Pull request by BishoyHanyRaafat**
https://drive.google.com/open?id=1PVJrobh4PflT3Vbg7HoEiDDH_jyxMNcN
https://drive.google.com/open?id=1EIywJT5rSXRE40TanCHWMAdGQpY6KBLu

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -52,6 +56,14 @@ To get started with the Pieces Python CLI Tool, you need to:
...
 After installing the CLI tool, you can access its functionalities through the terminal. The tool is initialized with the command `pieces` followed by various subcommands and options.
 
+
+### Some important terminologies
+
+- `x` -> The index
+- `current asset` -> The asset that you are currently using can be changed by the open command 
+- `current conversation` -> The conversation that you currently in which you will use in the ask command
+
+
```
Comment (hal-8999-alpha) [inline README.md:66]: Good stuff

File: README.md [hunk 2]
```diff
@@ -166,16 +178,69 @@ Change the model in the ask command.
...
+pieces conversation x
+```
+
+
 ##### Ask a question to a model:
-** Requires quotes around question **
+**Requires quotes around question**
 
-Ask the copoilt a question it uses chatGPT3 as a defualt model to ask a question, you can change the model using the change model command.
+Ask the copoilt a question it uses chatGPT3 as a defualt model to ask a question, you can change the model using the change model command. You can add a relevance file or snippet index check the list assets command.
```
Comment (hal-8999-alpha) [inline README.md:234]: Should be changed to GPT 3.5

File: README.md [hunk 3]
```diff
@@ -166,16 +178,69 @@ Change the model in the ask command.
...
+You can use `.` or `/` to refer to the current directory 
 
 ```bash
-pieces ask "your question"
+pieces ask "your question" -f /file1 /file2 folder -s 1 2 3
```
Comment (hal-8999-alpha) [inline README.md:238]: Interesting, so you're asking a question of a specific file?

File: src/pieces/api/config.py [hunk 4]
```diff
@@ -8,27 +8,32 @@
...
 # Defining the host is optional and defaults to http://localhost:1000
 # See configuration.py for a list of all supported configuration parameters.
 configuration = pos_client.Configuration(host=f"http://localhost:{port}")
 
 
+# Websocket config
+WEBSOCKET_URL = f"ws://localhost:{port}/qgpt/stream"
+TIMEOUT = 20  # seconds
+
 run_in_loop = False # is CLI looping?
```
Comment (hal-8999-alpha) [inline src/pieces/api/config.py:38]: Maybe add this to the issue regarding changing all the localhosts to the sdk

File: src/pieces/api/pieces_websocket.py [hunk 5]
```diff
@@ -95,48 +95,37 @@ def _start_ws(self):
...
             self.close_websocket_connection()
+        self.verbose = True
         if not finishes:
-            self.verbose = True
             raise ConnectionError("Failed to get the reponse back")
-        self.verbose = True
-        return self.final_answer
\ No newline at end of file
+        return self.final_answer
+
```
Comment (hal-8999-alpha) [inline src/pieces/api/pieces_websocket.py:131]: Thanks for cleaning this section up!

**Review by hal-8999-alpha — CHANGES_REQUESTED** [review]
This is an EXCELLENT addition. Thank you for this. I love the ability to see conversation history etc. Great work, just a couple minor questions in the comments and a request for an issue around the localhost vs SDK stuff.

File: README.md (hunk 3 above)
Comment (BishoyHanyRaafat) [reply README.md:238]: Or folder

Comment (hal-8999-alpha) [reply README.md:258]: Very cool

**Comment by BishoyHanyRaafat** [comment]
@hal-8999-alpha anything wrong with this PR?

**Comment by hal-8999-alpha** [comment]
> @hal-8999-alpha anything wrong with this PR?

Only minor thing I had was to change it to GPT 3.5 instead of ChatGPT3

File: src/pieces/commands/copilot/ask_command.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
+from pieces.commands import commands_functions
+from pieces.api.pieces_websocket import WebSocketManager
+from pieces_os_client import *
+from pieces.api.config import api_client
+import os
+from pieces.gui import show_error
+from pieces.api.assets import get_assets_info_list
+
+ws_manager = WebSocketManager()
+def ask(query, **kwargs):
```
Comment (caleb-at-pieces) [inline src/pieces/commands/copilot/ask_command.py:10]: good cleanup

File: src/pieces/commands/copilot/ask_command.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+    if files or snippets:
+        relevant = QGPTApi(api_client).relevance(
+            QGPTRelevanceInput(
+                            query=query,
+                            paths=files,
+                            assets=flattened_assets,
+                            application=commands_functions.application.id,
+                            model=commands_functions.model_id
+                        )).to_dict()
```
Comment (caleb-at-pieces) [inline src/pieces/commands/copilot/ask_command.py:47]: is it worth doing .to_dict() here? wouldn't this end up losing our intellisense?

**Review by caleb-at-pieces — CHANGES_REQUESTED** [review]
looking good 

just update that command description and LGTM!

File: src/pieces/api/config.py [hunk 8]
```diff
@@ -8,27 +8,32 @@
...
 # Check if the directory exists, if not, create it
 pieces_data_dir = user_data_dir(appauthor="pieces", appname="cli-agent",ensure_exists=True)
 
 applications_db_path = Path(
     pieces_data_dir,"applications.db"
 )  # path to our applications.db
 
-models_file  = Path(pieces_data_dir, "model_data.pkl") # model data file
+models_file  = Path(
+    pieces_data_dir, "model_data.pkl"
```
[the remark this answers is not in the document]
Comment (BishoyHanyRaafat) [reply src/pieces/api/config.py:19]: The current model that you are using

File: src/pieces/api/pieces_websocket.py [hunk 9]
```diff
@@ -72,21 +72,21 @@ def on_error(self, ws, error):
 
     def on_close(self, ws, close_status_code, close_msg):
         """Handle websocket closure."""
-        if self.verbose:
-            print("WebSocket closed")
+        # if self.verbose:
```
[the remark this answers is not in the document]
Comment (BishoyHanyRaafat) [reply src/pieces/api/pieces_websocket.py:75]: I felt they are useless so I commented them instead of deleting

**Review by caleb-at-pieces — APPROVED** [review]
good job

**Review by hal-8999-alpha — APPROVED** [review]

**Pull request merged**